src/server/container_utils.cc

```diff
@@ -23,6 +23,95 @@ extern "C" {
+// If no such key exists or a wrong type is found, the apropriate status is returned.
+// Optimized version of `FindFirstNonEmpty` below.
+OpResult<std::string> FindFirstNonEmptySingleShard(Transaction* trans, int req_obj_type,
+                                                   BlockingResultCb func) {
+  DCHECK_EQ(trans->GetUniqueShardCnt(), 1u);
+  std::string key;
+  auto cb = [&](Transaction* t, EngineShard* shard) -> Transaction::RunnableResult {
+    auto args = t->GetShardArgs(shard->shard_id());
+    auto ff_res = shard->db_slice().FindFirstReadOnly(t->GetDbContext(), args, req_obj_type);
+
+    if (ff_res == OpStatus::WRONG_TYPE)
+      return OpStatus::WRONG_TYPE;
+
+    if (ff_res == OpStatus::KEY_NOTFOUND)
+      return {OpStatus::KEY_NOTFOUND, Transaction::RunnableResult::AVOID_CONCLUDING};
+
+    CHECK(ff_res.ok());  // No other errors possible
+    ff_res->first->first.GetString(&key);
+    func(t, shard, key);
+    return OpStatus::OK;
+  };
+
+  // Schedule single hop and hopefully find a key, otherwise avoid concluding
+  OpStatus status = trans->ScheduleSingleHop(cb);
+  if (status == OpStatus::OK)
+    return key;
+  return status;
+}
+
+// Find first non-empty key (sorted by order in command arguments) and return it,
+// otherwise return not found or wrong type error.
+OpResult<ShardFFResult> FindFirstNonEmpty(Transaction* trans, int req_obj_type) {
+  DCHECK_GT(trans->GetUniqueShardCnt(), 1u);
+
+  using FFResult = std::tuple<PrimeKey, unsigned, ShardId>;  // key, argument index, sid
+  VLOG(2) << "FindFirst::Find " << trans->DebugId();
+
+  // Holds Find results: (iterator to a found key, and its index in the passed arguments).
+  // See DbSlice::FindFirst for more details.
+  std::vector<OpResult<FFResult>> find_res(shard_set->size());
+  std::fill(find_res.begin(), find_res.end(), OpStatus::KEY_NOTFOUND);
+
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    auto args = t->GetShardArgs(shard->shard_id());
+    auto ff_res = shard->db_slice().FindFirstReadOnly(t->GetDbContext(), args, req_obj_type);
+    if (ff_res) {
+      find_res[shard->shard_id()] =
+          FFResult{ff_res->first->first.AsRef(), ff_res->second, shard->shard_id()};
+    } else {
+      find_res[shard->shard_id()] = ff_res.status();
+    }
+    return OpStatus::OK;
+  };
+
+  trans->Execute(std::move(cb), false);
+
+  // If any key is of the wrong type, report it immediately
+  if (std::find(find_res.begin(), find_res.end(), OpStatus::WRONG_TYPE) != find_res.end())
+    return OpStatus::WRONG_TYPE;
+
+  auto comp = [trans](const OpResult<FFResult>& lhs, const OpResult<FFResult>& rhs) {
+    if (!lhs || !rhs)
+      return lhs.ok();  // push errors to back
+    size_t i1 = trans->ReverseArgIndex(std::get<ShardId>(*lhs), std::get<unsigned>(*lhs));
+    size_t i2 = trans->ReverseArgIndex(std::get<ShardId>(*rhs), std::get<unsigned>(*rhs));
+    return i1 < i2;
+  };
+
+  auto it = std::min_element(find_res.begin(), find_res.end(), comp);
```

Comment:
Can you please add the explanation of what min_element provides?
Also, what does it mean when returning `lhs.ok()` ? I understand you are saying that an element that is ok is smaller but what does it mean?

Reply:
Added a comment